File: src/test/java/vproxy/poc/ApplicationLevelHttpServer.java
package vproxy.poc;

import vclient.HttpClient;
import vclient.HttpResponse;
import vjson.JSON;
import vjson.util.ObjectBuilder;
import vjson.util.Transformer;
import vproxy.util.Logger;
import vproxy.util.Utils;
import vserver.RoutingContext;
import vserver.HttpServer;
import vserver.Tool;

import java.io.IOException;
import java.util.LinkedList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

public class ApplicationLevelHttpServer {
    public static void main(String[] args) throws Exception {
        System.out.println("Start server");
        new ApplicationLevelHttpServer().runServer();
        Thread.sleep(1_000);
        System.out.println("Start client to build initial data");
        new ApplicationLevelHttpServer().runClient();
    }

    private static class Service {
        final UUID id;
        final String name;
        String address;
        int port;

        private Service(String name, String address, int port) {
            this.id = UUID.randomUUID();
            this.name = name;
            this.address = address;
            this.port = port;
        }
    }

    private final List<Service> services = new LinkedList<>();
    private final Transformer tf = new Transformer()
        .addRule(Service.class, s -> new ObjectBuilder()
            .put("id", s.id.toString())
            .put("name", s.name)
            .put("ingressAddress", s.address)
            .put("ingressPort", s.port)
            .build());

    private void runClient() throws Exception {
        HttpClient client = HttpClient.to("127.0.0.1", 8080);
        client.pst("/api/v1/services").send(new ObjectBuilder()
                .put("name", "myservice1")
                .put("address", "192.168.0.1")
                .put("port", 80)
                .build(),
            (err, resp) -> printResponse("myservice1", err, resp));

        client.pst("/api/v1/services").send(new ObjectBuilder()
                .put("name", "myservice2")
                .put("address", "192.168.0.2")
                .put("port", 80)
                .build(),
            (err, resp) -> printResponse("myservice2", err, resp));

        Thread.sleep(500);
        client.get("/api/v1/services").send((err, resp) -> {
            if (err != null) {
                System.out.println("Request to fetch services failed: " + err);
                err.printStackTrace();
                return;
            }
            if (resp.status() != 200) {
                System.out.println("Request to fetch services failed: " + resp.bodyAsString());
                return;
            }
            JSON.Instance inst = resp.bodyAsJson();
            System.out.println("Fetch services result:");
            System.out.println(inst.pretty());
        });
    }

    private void printResponse(String name, IOException err, HttpResponse resp) {
        if (err != null) {
            System.out.println("Request failed for " + name + ": " + err);
            err.printStackTrace();
            return;
        }
        if (resp.status() != 200) {
            System.out.println("Request failed for " + name + ": " + resp.bodyAsString());
        } else {
            System.out.println("Request succeeded for " + name);
        }
    }

    private void runServer() throws Exception {
        HttpServer.create()
            .all("/api/v1/*", Tool.bodyJsonHandler())
            .get("/api/v1/services/:serviceId", this::getService)
            .get("/api/v1/services", this::listServices)
            .pst("/api/v1/services", this::createService)
            .put("/api/v1/services/:serviceId", this::updateService)
            .del("/api/v1/services/:serviceId", this::deleteService)
            .listen(8080);
    }

    private void listServices(RoutingContext rctx) {
        Logger.alert("listServices called");
        rctx.response().status(200).end(tf.transform(services));
    }

    private void createService(RoutingContext rctx) {
        JSON.Instance body = rctx.get(Tool.bodyJson);
        Logger.alert("listServices called with " + body);
        Service service;
        try {
            JSON.Object o = (JSON.Object) body;
            String name = o.getString("name");
            String address = o.getString("address");
            int port = o.getInt("port");

            if (!Utils.isIpLiteral(address)) {
                throw new IllegalArgumentException();
            }
            if (port < 1 || port > 65535) {
                throw new IllegalArgumentException();
            }

            service = new Service(name, address, port);
            services.add(service);
        } catch (RuntimeException e) {
            rctx.response().status(400)
                .end(new ObjectBuilder()
                    .put("message", "invalid request body")
                    .build());
            return;
        }
        rctx.response().status(200).end(tf.transform(service));
    }

    private void getService(RoutingContext rctx) {
        String serviceId = rctx.param("serviceId");
        Logger.alert("getService called with `" + serviceId + "`");
        Optional<Service> ret = services.stream().filter(s -> s.id.toString().equals(serviceId)).findAny();
        if (ret.isPresent()) {
            rctx.response().status(200).end(tf.transform(ret.get()));
        } else {
            rctx.response().status(404)
                .end(new ObjectBuilder()
                    .put("message", "service with id `" + serviceId + "` not found")
                    .build());
        }
    }

    private void updateService(RoutingContext rctx) {
        String serviceId = rctx.param("serviceId");
        JSON.Instance body = rctx.get(Tool.bodyJson);
        Logger.alert("updateService called with `" + serviceId + "` and " + body);

        String address = null;
        int port = -1;
        try {
            JSON.Object o = (JSON.Object) body;
            if (o.containsKey("address")) {
                address = o.getString("address");
                if (!Utils.isIpLiteral(address)) {
                    throw new IllegalArgumentException();
                }
            }
            if (o.containsKey("port")) {
                port = o.getInt("port");
                if (port < 1 || port > 65535) {
                    throw new IllegalArgumentException();
                }
            }
        } catch (RuntimeException e) {
            rctx.response().status(400)
                .end(new ObjectBuilder()
                    .put("message", "invalid request body")
                    .build());
            return;
        }

        Optional<Service> ret = services.stream().filter(s -> s.id.toString().equals(serviceId)).findAny();
        if (ret.isPresent()) {
            Service s = ret.get();
            if (address != null) {
                s.address = address;
            }
            if (port != -1) {
                s.port = port;
            }
            rctx.response().status(204).end();
        } else {
            rctx.response().status(404)
                .end(new ObjectBuilder()
                    .put("message", "service with id `" + serviceId + "` not found")
                    .build());
        }
    }

    private void deleteService(RoutingContext rctx) {
        String serviceId = rctx.param("serviceId");
        Logger.alert("deleteService called with `" + serviceId + "`");
        Optional<Service> ret = services.stream().filter(s -> s.id.toString().equals(serviceId)).findAny();
        if (ret.isPresent()) {
            services.remove(ret.get());
            rctx.response().status(204).end();
        } else {
            rctx.response().status(404)
                .end(new ObjectBuilder()
                    .put("message", "service with id `" + serviceId + "` not found")
                    .build());
        }
    }
}

File: test/src/test/kotlin/vproxy/poc/ApplicationLevelHttpServer.kt
package vproxy.poc

import vjson.JSON
import vjson.util.ObjectBuilder
import vjson.util.Transformer
import vproxy.base.connection.ConnectableConnection
import vproxy.base.connection.ServerSock
import vproxy.base.processor.http1.entity.Response
import vproxy.base.selector.SelectorEventLoop
import vproxy.base.util.ByteArray
import vproxy.base.util.Logger
import vproxy.base.util.kt.KT
import vproxy.base.util.thread.VProxyThread
import vproxy.lib.common.fitCoroutine
import vproxy.lib.common.launch
import vproxy.lib.common.sleep
import vproxy.lib.http.RoutingContext
import vproxy.lib.http.Tool
import vproxy.lib.http1.CoroutineHttp1Server
import vproxy.vfd.IP
import vproxy.vfd.IPPort
import java.util.*

@Suppress("BlockingMethodInNonBlockingContext")
class ApplicationLevelHttpServer {
  private class Service constructor(val name: String, var address: String, var port: Int) {
    val id: UUID = UUID.randomUUID()
  }

  private val services: MutableList<Service> = LinkedList<Service>()
  private val tf = Transformer()
    .addRule(KT.kclass(Service::class.java)) { s: Service ->
      ObjectBuilder()
        .put("id", s.id.toString())
        .put("name", s.name)
        .put("ingressAddress", s.address)
        .put("ingressPort", s.port)
        .build()
    }

  private fun runClient() {
    val loop = SelectorEventLoop.open()
    loop.ensureNetEventLoop()
    loop.loop { VProxyThread.create(it, "app-level-http-run-client") }
    val conn = ConnectableConnection.create(IPPort("127.0.0.1", 8080)).fitCoroutine(loop.ensureNetEventLoop())
    loop.launch {
      sleep(500)
      conn.connect()
      val http = conn.asHttp1ClientConnection()
      http.post("/api/v1/services").send(
        ObjectBuilder()
          .put("name", "myservice1")
          .put("address", "192.168.0.1")
          .put("port", 80)
          .build()
      )
      http.post("/api/v1/services").send(
        ObjectBuilder()
          .put("name", "myservice2")
          .put("address", "192.168.0.2")
          .put("port", 80)
          .build()
      )
      var resp = http.readResponse()
      printResponse("myservice1", resp)
      resp = http.readResponse()
      printResponse("myservice2", resp)

      http.get("/api/v1/services").send()
      resp = http.readResponse()
      if (resp.body == null) {
        resp.body = ByteArray.from("{}")
      }
      println("Fetch services result:")
      println(JSON.parse(resp.body.toString()))
    }.invokeOnCompletion {
      conn.close()
    }
  }

  private fun printResponse(name: String, resp: Response) {
    if (resp.statusCode != 200) {
      println("Request failed for " + name + ": " + resp.body)
    } else {
      println("Request succeeded for $name")
    }
  }

  private fun runServer() {
    val loop = SelectorEventLoop.open()
    loop.ensureNetEventLoop()
    loop.loop { VProxyThread.create(it, "app-level-http-run-server") }
    val svrsock = ServerSock.create(IPPort("::", 8080))
    val server = CoroutineHttp1Server(svrsock.fitCoroutine(loop.ensureNetEventLoop()))
    server
      .all("/*") { log(it) }
      .all("/api/v1/*", Tool.bodyJsonHandler())
      .get("/api/v1/services/:serviceId") { getService(it) }
      .get("/api/v1/services") { listServices(it) }
      .post("/api/v1/services") { createService(it) }
      .put("/api/v1/services/:serviceId") { updateService(it) }
      .del("/api/v1/services/:serviceId") { deleteService(it) }

    loop.launch {
      server.start()
    }
  }

  private fun log(rctx: RoutingContext) {
    Logger.alert(
      "received request remote=" + rctx.conn.base().remote.formatToIPPortString()
        .toString() + " -> local=" + rctx.conn.base().local.formatToIPPortString()
        .toString() + " " + rctx.req.method() + " " + rctx.req.uri()
    )
    rctx.allowNext()
  }

  private suspend fun listServices(rctx: RoutingContext) {
    Logger.alert("listServices called")
    rctx.conn.response(200).send(tf.transform(services))
  }

  private suspend fun createService(rctx: RoutingContext) {
    val body: JSON.Instance<*> = rctx.get(Tool.bodyJson)!!
    Logger.alert("listServices called with $body")
    val service: Service
    try {
      val o = body as JSON.Object
      val name = o.getString("name")
      val address = o.getString("address")
      val port = o.getInt("port")
      require(IP.isIpLiteral(address))
      require(!(port < 1 || port > 65535))
      service = Service(name, address, port)
      services.add(service)
    } catch (e: RuntimeException) {
      rctx.conn.response(400)
        .send(
          ObjectBuilder()
            .put("message", "invalid request body")
            .build()
        )
      return
    }
    rctx.conn.response(200).send(tf.transform(service))
  }

  private suspend fun getService(rctx: RoutingContext) {
    val serviceId: String = rctx.param("serviceId")
    Logger.alert("getService called with `$serviceId`")
    val ret = services.stream().filter { s: Service -> s.id.toString() == serviceId }.findAny()
    if (ret.isPresent) {
      rctx.conn.response(200).send(tf.transform(ret.get()))
    } else {
      rctx.conn.response(404)
        .send(
          ObjectBuilder()
            .put("message", "service with id `$serviceId` not found")
            .build()
        )
    }
  }

  private suspend fun updateService(rctx: RoutingContext) {
    val serviceId: String = rctx.param("serviceId")
    val body: JSON.Instance<*> = rctx.get(Tool.bodyJson)!!
    Logger.alert("updateService called with `$serviceId` and $body")
    var address: String? = null
    var port = -1
    try {
      val o = body as JSON.Object
      if (o.containsKey("address")) {
        address = o.getString("address")
        require(IP.isIpLiteral(address))
      }
      if (o.containsKey("port")) {
        port = o.getInt("port")
        require(!(port < 1 || port > 65535))
      }
    } catch (e: RuntimeException) {
      rctx.conn.response(400)
        .send(
          ObjectBuilder()
            .put("message", "invalid request body")
            .build()
        )
      return
    }
    val ret = services.stream().filter { s: Service -> s.id.toString() == serviceId }.findAny()
    if (ret.isPresent) {
      val s = ret.get()
      if (address != null) {
        s.address = address
      }
      if (port != -1) {
        s.port = port
      }
      rctx.conn.response(204).send()
    } else {
      rctx.conn.response(404)
        .send(
          ObjectBuilder()
            .put("message", "service with id `$serviceId` not found")
            .build()
        )
    }
  }

  private suspend fun deleteService(rctx: RoutingContext) {
    val serviceId: String = rctx.param("serviceId")
    Logger.alert("deleteService called with `$serviceId`")
    val ret = services.stream().filter { s: Service -> s.id.toString() == serviceId }.findAny()
    if (ret.isPresent) {
      services.remove(ret.get())
      rctx.conn.response(204).send()
    } else {
      rctx.conn.response(404)
        .send(
          ObjectBuilder()
            .put("message", "service with id `$serviceId` not found")
            .build()
        )
    }
  }

  companion object {
    @JvmStatic
    fun main(args: Array<String>) {
      println("Start server")
      ApplicationLevelHttpServer().runServer()
      Thread.sleep(1000)
      println("Start client to build initial data")
      ApplicationLevelHttpServer().runClient()
    }
  }
}
